### PR TITLE
update .lintr

### PR DESCRIPTION
**Pull request by MichaelChirico**

### CODE REVIEW

File: .lintr [hunk 1]
```diff
@@ -1,7 +1,7 @@
-linters: linters_with_defaults( # The following TODOs are part of an effort to have {lintr} lint-free (#584)
+linters: linters_with_defaults(
    line_length_linter = line_length_linter(120),
-   # TODO add implicit_integer_linter()
-   cyclocomp_linter = cyclocomp_linter(37) # TODO reduce to 15
+   implicit_integer_linter(),
+   cyclocomp_linter = cyclocomp_linter(37) # TODO(#1186) reduce to 15
```
Comment (AshesITR) [inline .lintr:4]: currently 40 actually...

**Comment by MichaelChirico** [comment]
well that had a lot bigger impact than expected :joy:

**Comment by AshesITR** [comment]
I wouldn't modify the default linter testcode.
Maybe `dummy_packages/clean` needs an empty `.lintr` config, so it doesn't pull our config?

**Comment by MichaelChirico** [comment]
there's already a config there:

https://github.com/r-lib/lintr/blob/5578732137ec174e44d1bc68a09e026eb49749b1/tests/testthat/dummy_packages/clean/.lintr

we need the exclusion in our top-level .lintr config don't we? since the annotation is being done by the GHA linter

**Comment by MichaelChirico** [comment]
oh but it's already there 🤔 not sure what's going on

https://github.com/r-lib/lintr/blob/5578732137ec174e44d1bc68a09e026eb49749b1/.lintr#L9

**Comment by AshesITR** [comment]
`lint-changed-files` doesn't pull the excludes I think, because it overrides them.

**Comment by AshesITR** [comment]
NB the test failures are because of lints in `tests/testthat/dummy_packages/clean` which makes no sense because there is a `.lintr` file there. Should investigate further.

**Comment by MichaelChirico** [comment]
still no luck btw... i'm also reproducing the failure locally now

File: .lintr_new [hunk 2]
```diff
@@ -1,13 +1,14 @@
 linters: linters_with_defaults(
     implicit_integer_linter(),
-    line_length_linter = line_length_linter(120)
+    line_length_linter(120)
   )
 exclusions: list(
+  "inst/doc/creating_linters.R" = 1,
+  "inst/example/bad.R",
```
[the remark this answers is not in the document]
Comment (AshesITR) [reply .lintr_new:7]: I'm not sure why they were different in the first place. Two states make sense to me: (a) exclusions are the same in `.lintr` and `.lintr_new` or (b) exclusions are empty in `.lintr_new`.

**Comment by AshesITR** [comment]
That's good, I don't.
No idea why the `.lintr` isn't found. Maybe something with the working directory has changed?
`find_config("dummy_packages/clean")` should find `"dummy_packages/clean/.lintr"`, no?

**Comment by MichaelChirico** [comment]
> That's good, I don't.

FWIW, `pkgload::load_all(); testthat::test_dir('tests')` isn't reproducing for me, but `R CMD build . && R CMD check lintr_2.0.1.9000.tar.gz` is

**Comment by AshesITR** [comment]
`devtools::load_all(); devtools::test()` and `devtools::load_all(); testthat::test_dir("tests")` work fine for me.
Our performance tuning really paid off though - getting 26s test time now, that's half of what it used to be 🚀

**Comment by MichaelChirico** [comment]
>work fine

that means it passes, or reproduces the failure?

> getting 26s test time now, that's half of what it used to be 🚀

bravo! should definitely mention performance improvement in the NEWS...

**Comment by AshesITR** [comment]
GHA builds start to go south after running `lint_package("dummy_packages/package", cache = TRUE)` twice.
Probably worth to start investigation there. Maybe something messes up the working directory there?

**Comment by AshesITR** [comment]
> > work fine
> 
> that means it passes, or reproduces the failure?

they pass, no failure.

> > getting 26s test time now, that's half of what it used to be rocket
> 
> bravo! should definitely mention performance improvement in the NEWS...

will do a PR to our news branch 😄

**Comment by MichaelChirico** [comment]
Simply refactoring the `test-defaults` test appears to have fixed it (?)

Still not quite sure what's going on with `expect_lint_free()` -- especially that the `.lintr` config exclusions are apparently ignored...

**Comment by MichaelChirico** [comment]
Manually setting `parse_settings=FALSE` here works for me locally (lets see if CI agrees)... I'd be happy to just merge this (was supposed to be a very simple PR :grimacing:) and file a follow-up bug

**Comment by AshesITR** [comment]
But `parse_settings = FALSE` basically circumvents the tests purpose, that local linter configs are loaded.

**Comment by MichaelChirico** [comment]
Let's disable the test with a TODO+bug then? Since we can't seem to reproduce the bug except inside `R CMD check`

**Comment by AshesITR** [comment]
I'll play around with it a bit. Maybe I find something.

**Comment by AshesITR** [comment]
Ah. The problem just struck me.
We `.Rbuildignore` the `.lintr` files needed for the tests!

**Comment by AshesITR** [comment]
If this works, I suggest we rename the `.lintr` files so `R CMD check` doesn't complain about hidden files and run all tests with `options(lintr.linter_file = "lintr_test_config")` instead.

**Comment by MichaelChirico** [comment]
oh great call!!! perfectly explains why the issue is only happening in R CMD check

**Review by AshesITR — APPROVED** [review]

**Pull request merged**